src/App.tsx

```diff
@@ -1,26 +1,254 @@
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setInputValue(event.target.value);
+  };
+
+  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (inputValue.trim() === '') {
+      setErrorMessage(ErrorMessages.EMPTY_TITLE);
+
+      return;
+    }
+
+    const temporaryTodo: Todo = {
+      id: 0,
+      title: inputValue.trim(),
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo(temporaryTodo);
+    setIsSubmitting(true);
+    setShouldFocus(false);
+    setLoadingTodosCount(currentCount => [...currentCount, 0]);
+
+    addTodo({ title: inputValue.trim(), userId: USER_ID, completed: false })
+      .then((createdTodo: Todo) => {
+        setTodos((currentTodos: Todo[]) => [...currentTodos, createdTodo]);
+        setInputValue('');
+        setErrorMessage(ErrorMessages.NO_ERRORS);
+        setShouldFocus(true);
+
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      })
+      .catch(() => {
+        setShouldFocus(true);
+        setErrorMessage(ErrorMessages.ADDING_ERROR);
+
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      })
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setIsSubmitting(false);
+        setTempTodo(null);
+        setLoadingTodosCount(currentCount =>
+          currentCount.filter(todoId => todoId !== 0),
+        );
+      });
+  };
+
+  const handleTodoDelete = useCallback(
+    (todoId: number) => {
+      setIsSubmitting(true);
+      setLoadingTodosCount(current => [...current, todoId]);
+      deleteTodos(todoId)
+        .then(() => {
+          setTodos(currentTodos =>
+            currentTodos.filter(todo => todo.id !== todoId),
+          );
+        })
+        .catch(() => {
+          setErrorMessage(ErrorMessages.DELETING_ERROR);
+        })
+        .finally(() => {
+          setIsSubmitting(false);
+          setLoadingTodosCount(current =>
+            current.filter(deletingTodoId => todoId !== deletingTodoId),
+          );
+        });
+    },
+    [todos],
+  );
+  const handleCompletedTodosDeleted = useCallback(() => {
+    setIsSubmitting(true);
+
+    const todosForDeleting = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setLoadingTodosCount(current => [...current, ...todosForDeleting]);
+
+    Promise.all(
+      todosForDeleting.map(id =>
+        deleteTodos(id)
+          .then(() => {
+            setTodos(currentTodos =>
+              currentTodos.filter(todo => todo.id !== id),
+            );
+          })
+          .catch(() => {
+            setErrorMessage(ErrorMessages.DELETING_ERROR);
+          })
+          .finally(() => {
+            setLoadingTodosCount(current =>
+              current.filter(deletingTodoId => deletingTodoId !== id),
+            );
```

Comment:
why are u changing the `loadingTodosCount` in `finally`, it will be counted not depending on the request status, maybe its better to count `loadingTodosCount` if the request ends with success.

Reply:
Deleting loadingTodosCount in the finally block is done to clear the loading indication for each todo, regardless of whether the delete operation was successful or failed.